python/rapidsmpf/rapidsmpf/integrations/single.py

```diff
@@ -0,0 +1,495 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Shuffler integration for single-worker pylibcudf execution."""
+
+from __future__ import annotations
+
+import threading
+from typing import TYPE_CHECKING, Any, cast
+
+import rmm.mr
+from rmm.pylibrmm.stream import DEFAULT_STREAM
+
+from rapidsmpf.buffer.buffer import MemoryType
+from rapidsmpf.buffer.resource import BufferResource, LimitAvailableMemory
+from rapidsmpf.communicator.single import new_communicator
+from rapidsmpf.config import Optional, OptionalBytes, Options
+from rapidsmpf.integrations.core import ShufflerIntegration, WorkerContext
+from rapidsmpf.progress_thread import ProgressThread
+from rapidsmpf.rmm_resource_adaptor import RmmResourceAdaptor
+from rapidsmpf.shuffler import Shuffler
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Sequence
+
+    from rapidsmpf.integrations.core import DataFrameT
+
+
+# Set of available shuffle IDs
+_shuffle_id_vacancy: set[int] = set(range(Shuffler.max_concurrent_shuffles))
+_shuffle_id_vacancy_lock: threading.Lock = threading.Lock()
+
+
+# Local single-worker context
+class _SingleWorker:
+    """Mutable single-worker utility class."""
+
+    context: WorkerContext | None = None
+
+
+_single_rapidsmpf_worker: _SingleWorker = _SingleWorker()
+
+
+def get_single_worker_context() -> WorkerContext:
+    """
+    Retrieve the single `WorkerContext`.
+
+    If the worker context does not already exist on the worker, it
+    will be created and attached to `_single_rapidsmpf_worker_context`.
+
+    Returns
+    -------
+    The existing or newly initialized worker context.
+    """
+    with WorkerContext.lock:
+        if _single_rapidsmpf_worker.context is None:
+            _single_rapidsmpf_worker.context = WorkerContext()
+        return cast(WorkerContext, _single_rapidsmpf_worker.context)
+
+
+def setup_single_worker(options: Options = Options()) -> None:
+    """
+    Attach RapidsMPF shuffling attributes to a single worker.
+
+    Parameters
+    ----------
+    options
+        Configuration options.
+
+    Warnings
+    --------
+    This function creates a new RMM memory pool, and
+    sets it as the current device resource.
+    """
+    ctx = get_single_worker_context()
+    with ctx.lock:
+        if ctx.comm is not None:
+            return  # Single worker already set up
+
+        ctx.options = options
+        ctx.comm = new_communicator(options)
+        ctx.comm.logger.trace("single communicator created.")
+
+        # Insert RMM resource adaptor on top of the current RMM resource stack.
+        mr = RmmResourceAdaptor(
```

Comment:
IIUC, there's some duplicate code here with https://github.com/rapidsai/rapidsmpf/blob/branch-25.08/python/rapidsmpf/rapidsmpf/integrations/dask/core.py#L199C5-L199C22. Any chance at sharing some of this?